### PR TITLE
libredirect: Fix redirects not working for subprocesses

### PR DESCRIPTION
**Pull request by nh2**
###### Motivation for this change

Fixes `libredirect` redirections not having any effect for subprocesses, because the environment variable was unset.

Found for example here: https://github.com/NixOS/nixpkgs/pull/124057#discussion_r637542270

The PR also enables debug symbols for better debugging of libredirect functionality.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Added a release notes entry if the change is major or breaking
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 124157` at 00c80aa1 run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li><a href="https://gist.github.com/83ce4107fc421851b9ae791d74679830">gns3-server</a></li>
  </ul>
</details>
<details>
  <summary>8 packages skipped due to time constraints:</summary>
  <ul>
    <li>citrix_workspace (citrix_workspace_21_03_0)</li>
    <li>citrix_workspace_20_04_0</li>
    <li>citrix_workspace_20_06_0</li>
    <li>citrix_workspace_20_09_0</li>
    <li>citrix_workspace_20_10_0</li>
    <li>citrix_workspace_20_12_0</li>
    <li>citrix_workspace_21_01_0</li>
    <li>home-assistant</li>
  </ul>
</details>
<details>
  <summary>28 packages built successfully:</summary>
  <ul>
    <li>alerta-server</li>
    <li>appdaemon</li>
    <li>cura</li>
    <li>gdbgui</li>
    <li>gns3-gui</li>
    <li>libredirect</li>
    <li>moodle-dl</li>
    <li>octoprint</li>
    <li>onionshare</li>
    <li>onionshare-gui</li>
    <li>python38Packages.aioambient</li>
    <li>python38Packages.cassandra-driver</li>
    <li>python38Packages.flask-socketio</li>
    <li>python38Packages.python-engineio</li>
    <li>python38Packages.python-engineio_3</li>
    <li>python38Packages.python-socketio</li>
    <li>python38Packages.python-socketio_4</li>
    <li>python38Packages.sentry-sdk</li>
    <li>python38Packages.simplisafe-python</li>
    <li>python39Packages.aioambient</li>
    <li>python39Packages.cassandra-driver</li>
    <li>python39Packages.flask-socketio</li>
    <li>python39Packages.python-engineio</li>
    <li>python39Packages.python-engineio_3</li>
    <li>python39Packages.python-socketio</li>
    <li>python39Packages.python-socketio_4</li>
    <li>python39Packages.sentry-sdk</li>
    <li>python39Packages.simplisafe-python</li>
  </ul>
</details>
<details>
  <summary>4 suggestions:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md">missing-phase-hooks</a>

`installPhase` should probably contain `runHook preInstall` and `runHook postInstall`.

Near pkgs/build-support/libredirect/default.nix:31:3:
```
   |
31 |   installPhase = ''
   |   ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/license-missing.md">license-missing</a>

Package is missing a license.

Near pkgs/build-support/libredirect/default.nix:54:3:
```
   |
54 |   meta = with lib; {
   |   ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md">missing-phase-hooks</a>

`buildPhase` should probably contain `runHook preBuild` and `runHook postBuild`.

Near pkgs/build-support/libredirect/default.nix:15:3:
```
   |
15 |   buildPhase = ''
   |   ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/maintainers-missing.md">maintainers-missing</a>

Package does not have a maintainer. Consider adding yourself?

Near pkgs/build-support/libredirect/default.nix:54:3:
```
   |
54 |   meta = with lib; {
   |   ^
```
</li>
  </ul>
</details>

Note that build failures [may predate](https://github.com/nix-community/hydra-check) this PR, and could be nondeterministic or hardware dependent.
Please exercise your independent judgement. Does something look off? Please file an [issue](https://github.com/rmcgibbo/r-rmcgibbo/issues) or reach out on IRC.

**Comment by nh2** [comment]
> 1 package failed to build:
> 
>     * [gns3-server](https://gist.github.com/83ce4107fc421851b9ae791d74679830)

Looks unrelated:

```
ERROR: Could not find a version that satisfies the requirement py-cpuinfo==7.0.0 (from gns3-server)
```

**Comment by nh2** [comment]
> warning: [missing-phase-hooks](https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md)

Fixed.

**Comment by jtojnar** [comment]
Would that be testable?

**Comment by nh2** [comment]
> Would that be testable?

@jtojnar I only tested so far that it fixes the use with `sublime4` / seems to have the intended purpose there.

**Comment by jtojnar** [comment]
See also https://github.com/NixOS/nixpkgs/issues/73077#issuecomment-662969771

**Comment by bachp** [comment]
@nh2 I run into the same issue that `libredirect`wasn't working in `cargo test`. This MR fixes the issue.

@jtojnar @edolstra Is there anything that prevents this from being merged?

**Comment by jtojnar** [comment]
I do not have objections but it would be nicer if there were tests.

**Comment by bachp** [comment]
@nh2 I rebase on current master and did a force push

@jtojnar I added a test that executes the test as a subprocess. I also verified that the test failed without this change.

**Review by bachp — APPROVED** [review]

File: pkgs/build-support/libredirect/test.c [hunk 1]
```diff
@@ -56,6 +61,14 @@ int main(void)
 
     test_spawn();
     test_system();
+
+    // Only run subprocess if no arguments are given
+    // as the subprocces will be calles without argument
```
Comment (jtojnar) [inline pkgs/build-support/libredirect/test.c:66]: subprocces → subprocess, calles → called

**Comment by risicle** [comment]
This broke libredirect on darwin.

**Comment by risicle** [comment]
Huh... although it _might_ (?) just be a hydra failure because it builds for me locally...

**Comment by NorfairKing** [comment]
I'd love to see this backported to 21.05.
It blocks integration tests that use Haskell's `mongoDB` library. (yes, really)
See https://github.com/NorfairKing/sydtest/commit/fce7178b30ad0db997bdbb4603966373e855b5c0

**Comment by nh2** [comment]
Anybody have objections to backporting it?

To me so far this counts as a bugfix. I will create the backport label, if you have objections, please comment on the PR that will be created by the backport bot.

**Comment by misuzu** [comment]
> Huh... although it _might_ (?) just be a hydra failure because it builds for me locally...

It consistently fails to build on my machine:
```
$ nix build nixpkgs#libredirect
error: builder for '/nix/store/m9hds9m5hyky38c10qy5zhvbni0azyxp-libredirect-0.drv' failed with exit code 134;
       last 10 log lines:
       > install: creating directory '/nix/store/s60rhw9d7kmzwn35l8hmdk5sllcyabv7-libredirect-0'
       > install: creating directory '/nix/store/s60rhw9d7kmzwn35l8hmdk5sllcyabv7-libredirect-0/lib'
       > 'libredirect.dylib' -> '/nix/store/s60rhw9d7kmzwn35l8hmdk5sllcyabv7-libredirect-0/lib/libredirect.dylib'
       > post-installation fixup
       > patching script interpreter paths in /nix/store/s60rhw9d7kmzwn35l8hmdk5sllcyabv7-libredirect-0
       > patching script interpreter paths in /nix/store/fdcbhajjqqpxckpj8vr48kb8sgp77qwc-libredirect-0-hook
       > running install tests
       > Assertion failed: (testfp != NULL), function main, file test.c, line 51.
       > Assertion failed: (system(SUBTEST) == 0), function test_subprocess, file test.c, line 41.
       > /nix/store/9mlkdg49j77i55v27jdl7kwcza5vvimk-stdenv-darwin/setup: line 1349: 90715 Abort trap: 6           NIX_REDIRECTS="/foo/bar/test=/nix/store/hf0xvn40qp9w14bf7cdkxvwsqa8c49q6-coreutils-8.32/bin/true" ./test
       For full logs, run 'nix log /nix/store/m9hds9m5hyky38c10qy5zhvbni0azyxp-libredirect-0.drv'.
$ nix shell nixpkgs#nix-info -c nix-info -m
 - system: `"x86_64-darwin"`
 - host os: `Darwin 20.6.0, macOS 10.16`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.4pre20210908_3c56f62`
 - channels(root): `"nixpkgs-21.11pre318361.a3a23d9599b"`
 - nixpkgs: `/nix/store/1y8fdj2a951by77n3vfpi6jyhmj2fvyg-source`
```

**Comment by nh2** [comment]
It did build on Darwin at the time of the PR according to Github Actions / Ofborg, so I'm not sure what to do about it.

**Pull request merged**